distributed-counter/src/main/java/org/infinispan/counter/CounterManager.java

```diff
@@ -0,0 +1,131 @@
+   }
+
+   /**
+    * It returns the {@link Counter} with that specific name.
+    * <p>
+    * If the {@link Counter} does not exists, it is created. Only the first invocation will take in consideration the
+    * initial value.
+    *
+    * @param name         the counter name.
+    * @param initialValue the counter initial value.
+    * @return the {@link Counter} instance.
+    */
+   public Counter getOrCreateCounter(String name, long initialValue) {
+      return counters.computeIfAbsent(name, counterName -> {
+         InfinispanCounter counter = new InfinispanCounter(counterName, cacheManager.getCache(COUNTER_CACHE_NAME));
+         counter.init(initialValue);
+         return counter;
+      });
+   }
+
+   /**
+    * It returns the {@link Counter} with that specific name.
+    * <p>
+    * If the {@link Counter} does not exists, it is created. Only the first invocation will take in consideration the
+    * initial value, min and max thresholds.
+    * <p>
+    * The initial value must be between the min and max value.
+    *
+    * @param name         the counter name.
+    * @param initialValue the counter initial value.
+    * @return the {@link Counter} instance.
+    * @throws IllegalArgumentException if the min, max or initial value are not valid.
+    */
+   public Counter getOrCreateCounter(String name, long initialValue, long minThreshold, long maxThreshold) {
+      if (minThreshold >= initialValue || initialValue >= maxThreshold) {
+         throw new IllegalArgumentException("Initial value must be between the min and max threshold.");
+      }
+      return counters.computeIfAbsent(name, counterName -> {
+         InfinispanCounter counter = new InfinispanCounter(counterName, cacheManager.getCache(COUNTER_CACHE_NAME));
+         counter.init(initialValue, minThreshold, maxThreshold);
+         return counter;
+      });
+   }
+
+   /**
+    * It returns the {@link Counter} with that specific name.
+    * <p>
+    * If the {@link Counter} does not exists, it is created. Only the first invocation will take in consideration the
+    * initial value.
+    * <p>
+    * The counter create is weakly consistent, meaning that the return value of {@link Counter#addAndGet(long)} (and
+    * similar) is meaningless and it should be used to be notified when the operation ended. Also, the {@link
+    * Counter#reset()} is not atomic and the intermediate result can be observed.
+    * <p>
+    * The most up to date value is only calculated via {@link Counter#getValue()} and it may not include in progress
+    * operations.
+    *
+    * @param name         the counter name.
+    * @param initialValue the counter initial value.
+    * @return the {@link Counter} instance.
+    */
+   public Counter getOrCreateWeakCounter(String name, long initialValue) {
+      return counters.computeIfAbsent(name, counterName -> {
+         WeakInfinispanCounter counter = new WeakInfinispanCounter(counterName, cacheManager.getCache(COUNTER_CACHE_NAME));
```

Comment:
I really hate when a method does not honor the contract on it's interface. Could we have `WeakCounter` interface where the affected methods would return `CompletableFuture<Void>`?